File: scanners/integrations/kms/base_kms_connector.py
"""
Base KMS Connector (Python) — Phase 14.3

Defines the abstract connector interface for cloud KMS, Vault, and HSM ecosystems.
Enforces read-only operation and metadata-only collection.
"""

from abc import ABC, abstractmethod
from typing import Dict, Any, List, Optional
from .kms_metadata import KmsKeyMetadata, assert_no_private_key_material


class BaseKmsConnector(ABC):
    """Abstract base class for KMS and HSM metadata connectors."""

    def __init__(
        self,
        name: str,
        provider: str,
        config: Optional[Dict[str, Any]] = None,
        read_only: bool = True,
        client: Any = None,
    ):
        if not name:
            raise ValueError("BaseKmsConnector requires 'name'")
        if not provider:
            raise ValueError("BaseKmsConnector requires 'provider'")

        # Invariant: Use least-privilege read-only roles by default
        if read_only is not True:
            raise PermissionError(
                f"Security Policy Violation: KMS connector '{name}' must be initialized in read-only mode (read_only=True)"
            )

        self.name = str(name)
        self.provider = str(provider).lower()
        self.config = dict(config or {})
        self.read_only = True
        self.client = client

        self.validate_config(self.config)

    def validate_config(self, config: Dict[str, Any]) -> None:
        """Validates provider-specific configuration."""
        if config is None:
            raise ValueError(f"{self.name} requires configuration dict")

    @abstractmethod
    def get_least_privilege_role_definition(self) -> Dict[str, Any]:
        """Returns the least-privilege read-only role or policy definition."""
        pass

    def validate_least_privilege(self, permissions: Any) -> Dict[str, Any]:
        """Validates that credentials/roles do not grant write or cryptographic operations."""
        return {"valid": True, "violations": []}

    def test_connection(self) -> Dict[str, Any]:
        """Performs a read-only probe to test connectivity."""
        return {"ok": True, "message": f"Read-only connection verified for {self.name}"}

    @abstractmethod
    def list_keys(self) -> List[Any]:
        """Lists key identifiers available in provider scope."""
        pass

    @abstractmethod
    def describe_key(self, key_id: str) -> KmsKeyMetadata:
        """Fetches metadata for a single key, returning KmsKeyMetadata."""
        pass

    def discover_all(self) -> List[KmsKeyMetadata]:
        """Discovers and describes all keys in scope."""
        key_items = self.list_keys()
        results = []

        for item in key_items:
            key_id = item if isinstance(item, str) else item.get("key_id") or item.get("id") or item.get("name")
            if not key_id:
                continue

            metadata = self.describe_key(key_id)
            if not isinstance(metadata, KmsKeyMetadata):
                raise TypeError(f"describe_key({key_id}) did not return a KmsKeyMetadata instance")

            assert_no_private_key_material(metadata.raw_metadata)
            results.push if hasattr(results, "push") else results.append(metadata)

        return results

    def sanitize_config(self) -> Dict[str, Any]:
        """Redacts secrets from connector configuration."""
        clean = dict(self.config)
        for k in clean.keys():
            if any(sub in k.lower() for sub in ["secret", "token", "password", "key", "auth", "credential"]):
                clean[k] = "[REDACTED]"
        return clean

File: backend/src/integrations/kms/base_kms_connector.js
/**
 * Base KMS & HSM Connector — Phase 14.3
 *
 * Vendor-agnostic connector abstraction for Cloud KMS, Vault, and HSM ecosystems.
 * Enforces least-privilege read-only operations and strict metadata-only collection.
 *
 * CRITICAL INVARIANTS:
 * 1. Metadata only (7 dimensions: keyId, algorithm, size, state, rotation, owner, usage).
 * 2. Never extract protected private key material.
 * 3. Use least-privilege read-only roles by default.
 */

const { KmsKeyMetadata, ProtectedKeyMaterialError } = require("./kms_metadata");

class BaseKmsConnector {
  /**
   * @param {Object} options
   * @param {string} options.name - Instance identifier
   * @param {string} options.provider - 'aws_kms' | 'gcp_kms' | 'azure_keyvault' | 'hashicorp_vault' | 'pkcs11_hsm'
   * @param {Object} [options.config] - Connector configuration (region, endpoint, credentials)
   * @param {boolean} [options.readOnly=true] - Enforce read-only mode (must be true)
   * @param {Object} [options.client] - Pluggable SDK / API client for testing/mocking
   * @param {Function} [options.fetchFn] - HTTP fetch function
   */
  constructor({
    name,
    provider,
    config = {},
    readOnly = true,
    client = null,
    fetchFn = globalThis.fetch,
  } = {}) {
    if (!name) throw new Error("BaseKmsConnector requires 'name'");
    if (!provider) throw new Error("BaseKmsConnector requires 'provider'");

    // Invariant: Use least-privilege read-only roles by default
    if (readOnly !== true) {
      throw new Error(
        `Security Policy Violation: KMS connector '${name}' must be initialized in read-only mode (readOnly: true)`
      );
    }

    this.name = String(name);
    this.provider = String(provider).toLowerCase();
    this.config = Object.freeze({ ...config });
    this.readOnly = true;
    this.client = client;
    this.fetchFn = fetchFn || globalThis.fetch;

    this.validateConfig(this.config);
  }

  /**
   * Validates vendor-specific configuration. Subclasses should override.
   * @param {Object} config
   */
  validateConfig(config) {
    if (!config) throw new Error(`${this.name} requires configuration object`);
  }

  /**
   * Returns the minimum least-privilege read-only IAM/role policy required.
   * @returns {Object} Policy definition and description
   */
  getLeastPrivilegeRoleDefinition() {
    throw new Error(`getLeastPrivilegeRoleDefinition() must be implemented by ${this.constructor.name}`);
  }

  /**
   * Validates that the active credential/role does not grant write/decrypt permissions.
   * @param {Object} [roleOrPermissions]
   * @returns {{ valid: boolean, violations: string[] }}
   */
  validateLeastPrivilege(roleOrPermissions = {}) {
    return { valid: true, violations: [] };
  }

  /**
   * Tests connectivity to the KMS/HSM service using read-only probes.
   * @returns {Promise<{ ok: boolean, message: string, details?: Object }>}
   */
  async testConnection() {
    return { ok: true, message: `Read-only connection test succeeded for ${this.name}` };
  }

  /**
   * Lists key identifiers available in the KMS/HSM scope.
   * @returns {Promise<Array<string|Object>>}
   */
  async listKeys() {
    throw new Error(`listKeys() must be implemented by ${this.constructor.name}`);
  }

  /**
   * Fetches metadata for a specific key.
   * @param {string} keyId
   * @returns {Promise<KmsKeyMetadata>}
   */
  async describeKey(keyId) {
    throw new Error(`describeKey() must be implemented by ${this.constructor.name}`);
  }

  /**
   * Discovers and returns metadata for all keys in scope.
   * Automatically asserts zero protected private key material.
   * @returns {Promise<Array<KmsKeyMetadata>>}
   */
  async discoverAll() {
    const keyItems = await this.listKeys();
    const results = [];

    for (const item of keyItems) {
      const keyId = typeof item === "string" ? item : item.keyId || item.id || item.name;
      if (!keyId) continue;

      const metadata = await this.describeKey(keyId);
      if (!(metadata instanceof KmsKeyMetadata)) {
        throw new Error(`describeKey(${keyId}) did not return a KmsKeyMetadata instance`);
      }

      // Explicit invariant guard
      KmsKeyMetadata.assertNoPrivateKeyMaterial(metadata.rawMetadata);
      results.push(metadata);
    }

    return results;
  }

  /**
   * Helper to sanitize configuration or headers for logging/output.
   */
  sanitizeConfig(cfg = this.config) {
    const clean = { ...cfg };
    for (const key of Object.keys(clean)) {
      if (/(secret|token|password|key|auth|credential)/i.test(key)) {
        clean[key] = "[REDACTED]";
      }
    }
    return clean;
  }
}

module.exports = {
  BaseKmsConnector,
};
